### ISSUE TITLE
Spec: In-Browser AI Depth Map Generator (Depth Anything V2)

### ISSUE THREAD

**Issue by manuelhe** [fractalsoftware/stereogramer#10]
## Problem Statement

Users of Stereogramer Studio want to generate captivating 3D autostereograms (SIRDS and Textured SIS) from everyday 2D images—such as family portraits, pets, landscape photography, digital illustrations, or product shots. However, creating a 3D stereogram requires a calibrated Depth Map where pixel values represent continuous distances in a 3D scene.

Currently, users must either rely on simple built-in geometric shapes/extruded text or use external 3D authoring software (e.g. Blender, Maya) and LiDAR/TrueDepth depth sensors to manually produce a grayscale depth map before uploading it. This friction prevents everyday creators, casual users, and artists without 3D modeling expertise from transforming their favorite photographs into autostereograms directly inside the web studio.

## Solution

Integrate an in-browser, client-side AI Monocular Depth Estimation pipeline directly into Stereogramer Studio. 

Powered by quantized lightweight neural network weights (Depth Anything V2 Small) running via ONNX Runtime Web and WebGPU/WASM in a dedicated background Web Worker, users can upload or drop any standard 2D image (JPEG, PNG, WebP) into the studio. The system automatically computes a high-resolution, continuous Depth Map in seconds—completely client-side, with zero server compute, zero API keys, and 100% user data privacy.

The estimated Depth Map seamlessly plugs into the core autostereogram pipeline, enabling immediate live preview, Depth Polarity inversion, Gaussian smoothing to prevent perceptual tearing, Bevel Extrusion, Disparity tuning, and high-resolution export in both parallel and cross-eyed convergence modes.

## User Stories

1. As a photographer, I want to drop a standard 2D portrait photo into the studio, so that the application automatically estimates depth and converts my photo into a 3D autostereogram without manual 3D modeling.
2. As a pet owner, I want to upload a photo of my cat or dog and have its organic contours automatically elevated in 3D, so that I can create a fun custom stereogram gift.
3. As an artist, I want to upload digital sketches and paintings to obtain an estimated Depth Map, so that my 2D artwork is transformed into an interactive autostereogram.
4. As a privacy-conscious user, I want the depth estimation model to execute entirely client-side in my browser, so that my personal photos never leave my device or get uploaded to a cloud server.
5. As a web user on a modern browser with WebGPU support, I want the depth model to execute on my GPU, so that inference takes only a few hundred milliseconds.
6. As a user on a system without WebGPU (or an older browser), I want the system to seamlessly fall back to multi-threaded WebAssembly (WASM) CPU inference, so that depth estimation works reliably across all supported platforms.
7. As a web visitor on first use, I want clear visual progress feedback (downloading model weights, percentage, initializing runtime, processing image), so that I know the system is actively working and not frozen.
8. As a repeat visitor, I want the downloaded ONNX model weights to be cached persistently in browser storage (Cache API / IndexedDB), so that subsequent visits don't re-download model weights over the network.
9. As a mobile or low-bandwidth user, I want the model weights to be compact and quantized (under 40 MB), so that downloading the model does not consume excessive data.
10. As a studio user, I want an "AI 2D-to-Depth" option in the Depth Map Source tab bar, so that I can easily discover and activate the feature alongside Presets, Shapes, Text, and Upload.
11. As a user who uploads a standard image in the existing Upload dropzone, I want a toggle switch "Estimate 3D Depth (AI)" with an explanatory tooltip, so that I can choose whether an uploaded image is already a pre-rendered grayscale depth map or a 2D photo requiring depth estimation.
12. As a creator, I want to see the estimated Depth Map preview displayed in the Reference Previews panel alongside the resulting Autostereogram, so that I can visually verify how depth values correspond to the original image features.
13. As a user, I want an Invert Depth Polarity toggle that updates the estimated depth map in real time, so that I can flip foreground and background if the AI model or visual composition feels inverted.
14. As a user, I want to adjust Bevel Extrusion and Gaussian Blur sliders on top of the AI-generated Depth Map, so that sharp Depth Discontinuities are smoothed out to prevent visual tearing and ocular fusion strain.
15. As a viewer, I want to tune Pattern Separation and Relief Depth Factor on the AI-generated depth map, so that the 3D depth pop is comfortable and easily fusible for my eyes.
16. As an autostereogram creator, I want to pair the AI-generated Depth Map with either Textured SIS (using custom pattern tiles or procedural textures) or SIRDS (noise palettes), so that I have complete creative freedom over the visual substrate.
17. As an interactive user, I want depth estimation to execute in a separate background Web Worker, so that the main UI thread never stutters, locks up, or drops frames while the model is computing.
18. As a user who changes their mind, I want an easy way to cancel an in-progress depth inference or remove the generated depth map and revert back to a preset or shape.
19. As a user with high-resolution photos, I want large images to be appropriately downscaled for model inference (e.g. 518×518 input dimension) and bilinearly upsampled back to the target stereogram canvas dimensions, so that processing remains fast and memory-safe without causing browser tab crashes.
20. As a user downloading the finished artwork, I want to export high-resolution PNG or JPEG files encoding the stereogram generated from the AI depth map, so that I can print or share it.
21. As a keyboard and accessibility user, I want accessible ARIA labels, focus states, and screen-reader announcements describing the model loading and inference state transitions.

### 1. Model Architecture & Runtime
- **Model Selection**: Deploy **Depth Anything V2 Small** quantized ONNX weights (`onnx-community/Depth-Anything-V2-Small-ONNX`). This model provides state-of-the-art monocular depth fidelity across arbitrary indoor/outdoor scenes, natural objects, and humans with a compact quantized footprint (~25MB to 35MB).
- **Execution Library**: Use `@huggingface/transformers.js` (v3) built on top of `onnxruntime-web`.
- **Hardware Acceleration**: Configure the pipeline to request the `webgpu` execution provider first; if WebGPU is not supported by the client device/browser or context creation fails, gracefully fall back to the multi-threaded `wasm` CPU execution provider.
- **Worker Isolation**: The entire ONNX runtime, model loading, image tensor transformation, and inference pipeline must live inside a dedicated Web Worker (`depth-estimation.worker.ts`) separate from the existing autostereogram generation worker. This prevents heavy tensor allocations or WASM execution from blocking UI interaction or canvas repainting.

### 2. Core Engine Decoupling (Adherence to ADR 0002)
- In accordance with ADR 0002 (Zero-Dependency Pure TypedArray Core Engine), `@stereogramer/core` will remain completely free of external dependencies, ML libraries, or ONNX bindings.
- The depth estimation worker converts the neural network tensor output into the core engine's standard domain representation: a normalized `Float32Array` elevation grid wrapped in a `DepthMap` descriptor:
  ```ts
  interface DepthMap {
    width: number;
    height: number;
    data: Float32Array; // values normalized between 0.0 (far background) and 1.0 (near foreground)
  }
  ```
- The resulting `DepthMap` is transferred across thread boundaries via transferable `ArrayBuffer` payloads and consumed by `@stereogramer/core` exactly like any procedural or uploaded depth map.

### 3. Studio UI/UX Integration (`packages/web`)
- **Depth Map Source Tab Bar**:
  - Add a new tab `AI Photo` (or `2D Photo`) to the existing `Depth Map Source` selector (`Presets`, `Shapes`, `Text`, `Upload`, `AI Photo`).
  - Alternatively, inside the `Upload` tab, introduce a segmented control: `Grayscale Depth Map` vs. `Estimate 2D Photo (AI)`.
- **Progress & Loading State**:
  - Display non-intrusive progress states: "Downloading AI Model (X%)...", "Compiling Shaders / Initializing WebGPU...", and "Estimating 3D Depth...".
  - Subsequent runs skip the download step as weights are persistently stored in browser cache.
- **Image Preprocessing & Normalization**:
  - Uploaded images of arbitrary resolution are loaded into an offscreen canvas, scaled to the model's preferred square patch grid (e.g. 518×518), and processed.
  - Model disparity logits are min-max normalized into $[0.0, 1.0]$ where $1.0$ represents the nearest foreground feature and $0.0$ represents the furthest background feature.
  - The normalized depth field is bilinearly resampled to match the active stereogram canvas dimensions ($640 \times 480$ or user dimensions).
- **Post-Processing Synergy**:
  - The existing Bevel Extrusion, Gaussian Blur kernel, and Depth Inversion toggles operate seamlessly on top of the estimated depth map before passing it to the TIW constraint solver.

### 4. Memory & Resource Safety
- All intermediate tensor buffers and WebGL/WebGPU textures are explicitly disposed of following inference to avoid browser memory leaks.
- If model download fails (e.g. offline or network disruption) or inference throws an out-of-memory error, a friendly user alert is surfaced with an automatic option to retry or fall back to procedural depth presets.

### What Makes a Good Test
- Tests must verify external observable behavior (UI state transitions, worker message protocols, depth normalization invariant $0.0 \le Z \le 1.0$, and canvas rendering updates) rather than internal neural network layer weights or ONNX internals.
- End-to-end tests must be resilient to headless CI environments where WebGPU may be unavailable by verifying that the fallback mechanisms and mocked worker pipelines function predictably.

### Modules Tested
- **Depth Estimation Worker / Pipeline Unit Tests (`packages/web`)**:
  - Test the tensor-to-DepthMap conversion and min-max normalization logic with synthetic 2D image matrices.
  - Test that output depth buffers strictly adhere to the $[0.0, 1.0]$ bounds invariant.
  - Test error handling when given corrupted image data or degenerate dimensions.
- **Web Studio E2E Integration Tests (Playwright in `packages/web/e2e/`)**:
  - Verify switching to the AI Photo depth source tab.
  - Verify that dropping a fixture image triggers the progress state and completes with an updated stereogram canvas.
  - Verify that toggling "Invert Depth Polarity" flips the depth values of the AI-generated map and immediately updates the autostereogram preview.
  - Verify that the layout remains side-by-side without visual regression.

### Prior Art
- Existing Playwright E2E suite (`packages/web/e2e/studio.spec.ts`) testing preset loading, drawer navigation, slider scrubbing, and canvas exports.
- Vitest unit tests in `@stereogramer/core` verifying TypedArray invariants and boundary conditions.

## Out of Scope
- Server-side cloud inference or hosting Python/PyTorch backends (all processing is 100% client-side in the browser).
- Multi-view 3D mesh generation or 3D Gaussian Splatting (this spec targets monocular depth maps for 2D autostereograms).
- Real-time video/webcam live stream depth estimation (single-image photo inference only).
- Fine-tuning or retraining model checkpoints (uses pre-trained quantized checkpoints).

## Further Notes
- Fully conforms to the domain glossary in `CONTEXT.md`.
- Strictly preserves ADR 0001 (TIW with HSR) and ADR 0002 (Zero-dependency TypedArray core engine).
